### ISSUE TITLE
Submodule repository unable to find working directory

### ISSUE THREAD

**Issue by logancollins** [libgit2/libgit2#4242]
Preface: Forgive me if this is expected behavior. I asked a question on Stack Overflow a couple weeks ago in the libgit2 tag, but never got a response. (https://stackoverflow.com/questions/43530123/libgit2-determining-submodule-working-directory-paths-with-git-repository-workd)

I'm noticing a behavior with submodule repositories (opened either directly or with `git_submodule_open`) where the results of `git_repository_workdir` is technically correct, but not specifically helpful.

For a main repository at `/MyProject`, and a submodule at `/MyProject/MySubmodule`, the submodule repository is opening with a workdir of `MySubmodule/`. However, since the actual `.git` directory for that repository (as well as the commondir) is at `/MyProject/.git/modules/MySubmodule`, libgit2 attempts to resolve working directory operations at `/MyProject/.git/modules/MySubmodule/MySubmodule` because the common dir is the same as the repository dir.

I can get around this by manually setting the workdir (using `git_repository_set_workdir`), but there is currently no exact way to determine the correct working directory without 1.) opening the parent repository and rebuilding the full submodule workdir's path manually or 2.) opening the submodule's configuration and checking to see if `core.worktree` is set (which it often is).

It doesn't appear libgit2 attempts to do either of these automatically, which leads me to think it's not the _proper_ way to do it.

### Reproduction steps

Open a submodule repository and attempt to perform a Diff from the index to working directory. It will fail with a `"cannot open directory /MySubmodule"` `GIT_ENOTFOUND` error.

### Expected behavior

The submodule repository should either have a workdir set that isn't relative to its repository (since it's a worktree), or there should be some indication of how it needs to resolve the path correctly.

### Actual behavior

The workdir cannot be resolved by operations that act on the working directory, such as Diff, Status Lists, etc.

### Version of libgit2 (release number or SHA1)

libgit2 `87f5fba` (has been reproducing for a couple of weeks worth of updates at least)

### Operating system(s) tested

macOS 10.12

**Comment by pks-t** [comment]
I'm unable to reproduce (see #4243 for my reproduction attempt where I open the submodules directly and via `git_submodule_open`, but paths are correct in both cases. Can you maybe produce a minimal working example? If not, could you please provide your gitmodules file?

**Comment by logancollins** [comment]
Of course! Sorry, that should've been the first thing I did.

In the example attached, there's one submodule (ExampleSubmodule) that reproduces the issue. The superproject itself is fine.

[ExampleProject.zip](https://github.com/libgit2/libgit2/files/1008441/ExampleProject.zip)

Here is some output of the libgit2 functions for that submodule:

```p git_repository_path(submoduleRepository)
"/Users/logan/Code/ExampleProject/.git/modules/ExampleSubmodule/"

p git_repository_commondir(submoduleRepository)
"/Users/logan/Code/ExampleProject/.git/modules/ExampleSubmodule/"

p git_repository_workdir(submoduleRepository)
"./"```

In this example I made, strangely, the workdir returned is not the path in the parent's folder, but just a reference to "./". So, instead of diffs and status lists being unable to build, they report that all files in the submodule are in a deleted state.

**Comment by pks-t** [comment]
Thanks for the repository, I can reproduce the error with it.

Problem is that your submodule contains a `gitdir` file at `ExampleProject/.git/modules/ExampleSubmodule/gitdir`, which should only exist for worktrees. We _could_ fix this issue if the `gitdir`file actually contained a sensible value, but its contents are simply ".git". Taking a look at gitrepository-layout(5), this file is only ever allowed to contain an absolute path.

I was able to work around the issue by making our checks whether a repository is a worktree or not more strict, which I'll add to #4243 later this day. But actually, you should try to get your tooling fixed. Do you have any idea how you ended up in this state? Did you maybe experiment with git-worktree(1) or manually moved some files around?

**Comment by logancollins** [comment]
Not that I know of. These two repos were made on macOS 10.12, checked into our self-hosted instance of GitLab, and the the submodule was added to the super project which cloned it into place. So maybe something with GitLab? I will try and investigate a bit more later today.

**Comment by pks-t** [comment]
Could you please test #4243 to see whether that fixes your problem?

**Comment by logancollins** [comment]
Yes! It appears this fixes it. The submodule repository finds its workdir properly now and no longer reports that it is a worktree.

A bit more investigative information if it's helpful at all:

This appears to be caused by a bug in older versions of Git. I was running Git 2.6.2 installed via Homebrew, and all submodule clones were getting this relative `gitdir` file. After updating to 2.13 the `gitdir` files are no longer being created.

Sorry for the red herring, but thank you so much for adding additional checks to prevent repositories that already have this issue from continuing to have issues.

**Comment by pks-t** [comment]
Okay, interesting to know and thanks for the hint and testing. :)

**Comment by ethomson** [comment]
Fixed via #4243
